[라면] API 테스트/문서자동화 미션 제출합니다.

Ramen6315 opened the pull request:
감사합니다.

Ramen6315 committed:
docs : README.md 파일 작성

Commit:
docs: 1단계 README.md 작성

Ramen6315 committed:
feat: 회원 가입 화면 연결

Commit:
setting: Basic, SessionInterceptor 제거

Commit:
feat: 회원 가입 API 연결

Ramen6315 committed:
setting: Gson Gradle 추가

Ramen6315 committed:
test: 회원 가입 성공 테스트 구현

Commit:
test: 회원 가입 유효성 검증 테스트

Ramen6315 committed:
refactor: MemberRequest field 어노테이션 수정

Ramen6315 committed:
feat: api 문서 자동화 구현

Commit:
test: JwtTokenProviderTest 작성

Ramen6315 committed:
feat: 로그인 기능 구현

Commit:
feat: Login 이메일 비밀번호 유효성 확인

Commit:
feat: Login 이메일 비밀번호 포맷 유효성 확인

Commit:
feat: Login 이메일 비밀번호 포맷 유효성 확인

Commit:
feat: 더미 데이터 삽입 ApplicationRunner

Ramen6315 committed:
test: 로그인 테스트 코드 작성

Commit:
setting: test application.properties 생성

Commit:
feat: Login Rest docs 생성

Ramen6315 committed:
feat: session 저장소에 token 저장

Ramen6315 committed:
test: Service Test 작성

Commit:
docs: 회원 정보 조회,수정,삭제 기능 목록 작성

Commit:
feat: 회원 정보 조회, 수정, 삭제 화면 연결

Ramen6315 committed:
test: 정보 조회 테스트 구현

Ramen6315 committed:
feat: BearerAuthInterceptor preHandle 구현

Ramen6315 committed:
refactor: InvalidAuthenticationException 생성자 수정

Ramen6315 committed:
refactor: 불필요 주석 삭제

Commit:
fix: 회원가입시 토큰 검증 불필요 로직 분기

BearerAuthInterceptor /Members 로 접근하는 모든 요청을
Interceptor에서 토큰 유효성 검증을 하여 발생하는 에러를
POST일 경우 건너뛰게 변경

Ramen6315 committed:
test: 회원 정보 수정 테스트 구현

Commit:
refactor: 세션에 email을 저장하도록 변경

 - Login.js 세션에 email을 저장하도록 변경
 - LoginMemberController 로그인 반환 DTO 변경

Commit:
feat: 마이 페이지 정보 조회 API 연결

Commit:
refactor: updateMember 메서드 반환값 변경

Commit:
feat: UpdateMember argument resolver 추가

Commit:
feat: member 이메일 동등성 확인 메서드 추가

Commit:
feat: update API 권한 확인 로직 추가

Commit:
fix: 회원 정보 가져오기

 - LoginMemberMethodArgumentResolver session에 저장된
 키값의 이름 변경

Commit:
feat: 내 정보 수정 API 연결

Commit:
feat: 회원 삭제 기능 구현 및 테스트 코드 작성

Commit:
fix: 회원 수정 로직 변경

mock 메서드 추가

Commit:
fix: Controller Advice 적용 시 발생 테스트 수정

Commit:
feat: 회원 삭제 기능 API 연결

Commit:
feat: README.md 2단계 즐겨찾기 기능목록 작성

Commit:
feat: FAVORITE table 추가

Commit:
setting: front sample file 추가

Commit:
feat: 즐겨 찾기 인수 테스트 작성

Commit:
feat: 즐겨 찾기 추가 테스트 작성

Commit:
feat: 즐겨 찾기 추가 기능 구현

Commit:
feat: Favorite 추가 인증 기능 구현

Ramen6315 committed:
feat: 즐겨 찾기 추가 Controller Test작성 및 구현

Ramen6315 committed:
test: 회원이 아닌 이메일 JWT 토큰 헤더 요청 시 즐겨찾기 추가 요청 테스트 변경

Ramen6315 committed:
feat: 즐겨찾기 추가 API 연결

Ramen6315 committed:
test: 즐겨 찾기 조회 테스트 코드 작성

Ramen6315 committed:
feat: Controller, Service, 예외 처리 테스트 코드 작성

Ramen6315 committed:
feat: snackbar.js 추가

Ramen6315 committed:
setting: Favorite 더미데이터 Configuration 추가

Ramen6315 committed:
fix: member 조회 NPE 해결

Ramen6315 committed:
refactor: FavoriteResponse ID 추가

개별적인 Favorite 요청을 식별하기 위해 ID 값을 DTO에 추가

Ramen6315 committed:
feat: Favorite 조회 API 연결 및 뷰 생성

Ramen6315 committed:
refactor: FavoriteDeleteAPI가 Id를 받도록 테스트 코드 수정

 - 데이터를 식별 할 수 있도록 Id로 Delete 변경

Ramen6315 committed:
refactor: FavoriteDeleteAPI 가 Id를 요청 받도록 Production code 변경

Ramen6315 committed:
refactor: FavoriteDeleteAPI 뷰 연결

Ramen6315 committed:
feat: README 파일 갱신

Ramen6315 committed:
feat: 회원탈퇴 api docs 추가

Ramen6315 committed:
feat: 회원 수정 api docs 추가

Ramen6315 committed:
feat: 회원 조회 api docs 추가

Ramen6315 committed:
feat: 즐겨 찾기 조회 api docs 추가

Ramen6315 committed:
feat: 즐겨 찾기 추가 api docs 추가

Ramen6315 committed:
feat: 즐겨 찾기 삭제 api docs 추가

Ramen6315 committed:
refactor: 코드 수정

Ramen6315 committed:
fix: 멤버 인수 테스트 수정

 - 헤더를 받지 않던 요청에 헤더 추가

Ramen6315 committed:
Merge pull request #1 from slowCoyle/step1-coyle

페어 프로젝트 머지

Hue9010 commented on src/main/resources/schema.sql:
```diff
@@ -38,4 +38,14 @@ create table if not exists MEMBER
+    primary key(id)
+--     foreign key(member_email) REFERENCES MEMBER (email) ON DELETE CASCADE ON UPDATE CASCADE,
+--     foreign key(source) REFERENCES STATION (name) ON DELETE CASCADE ON UPDATE CASCADE,
+--     foreign key(target) REFERENCES STATION (name) ON DELETE CASCADE ON UPDATE CASCADE
```
미사용 코드는 제거해주세요!

Hue9010 commented on src/main/java/wooteco/subway/web/FavoriteController.java:
```diff
@@ -0,0 +1,46 @@
+        this.favoriteService = favoriteService;
+    }
+
+    @PostMapping("/favorites")
```
RequestMapping을 클래스에 사용하여 공통 된 /favorites를 뺄 수 있겠네요.

Hue9010 commented on src/main/java/wooteco/subway/web/FavoriteController.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    @DeleteMapping("/favorites/{id}")
+    public ResponseEntity<Void> deleteFavorite(@CreateFavorite Favorite favorite) {
```
CreateFavorite 이름만 보면 생성 요청할 때 사용될 걸로 생각되네요.

Hue9010 commented on src/main/java/wooteco/subway/service/member/MemberService.java:
```diff
@@ -21,31 +23,36 @@ public Member createMember(Member member) {
         return memberRepository.save(member);
```
중복된 email로의 요청에 대한 처리도 추가해보세요.

Hue9010 commented on src/main/java/wooteco/subway/web/member/MemberController.java:
```diff
@@ -19,28 +22,27 @@ public MemberController(MemberService memberService) {
         return ResponseEntity.ok().body(MemberResponse.of(member));
     }
 
     @PutMapping("/members/{id}")
```
더이상 {id}는 사용안하네요.

Hue9010 commented on src/main/java/wooteco/subway/web/member/argumentresolver/UpdateMemberMethodArgumentResolver.java:
```diff
@@ -0,0 +1,41 @@
+        if (updateMember.isNotEqualEmail(email)) {
+            throw new UnAuthorizationException(email);
+        }
+        return updateMember;
```
이메일을 바꿀 수 없는건 권한의 문제인가요? 서비스쪽 코드를 보면 이메일에 대한 수정은 없는걸 보아 변경 불가한 데이터로 설계하신 걸로 보입니다. 그렇다면 이메일 변경은 인증/인가와는 별개가 아닐까 생각되는데 이 부분은 라면이 설계한 회원 수정 정책이 어떨지 알아야 할 것 같네요.
혹은 필요에 의해 만들진게 아니라 리졸버를 추가하는 걸 먼저 확정 한 다음 사용해 볼 곳을 찾은게 아닐까 추측되네요. 리졸버를 만들어보는 것에 목적을 뒀다면 잘 만들어 주셨습니다. 👍  그게아니라면 정말 필요한가 의문을 가져 볼 필요가 있습니다. 
결론적으로는 학습을 위해 의도적으로 과도하게 사용했는지 아닌지가 큰 차이를 가집니다.
(사용안해보면 알기 어려우니 사용해보는 걸 장려하긴 합니다. 😉 )

Hue9010 commented on src/main/java/wooteco/subway/web/member/interceptor/BearerAuthMemberInterceptor.java:
```diff
@@ -0,0 +1,95 @@
+        String email = jwtTokenProvider.getSubject(bearer);
+        request.setAttribute("requestMemberEmail", email);
+
+        if (isGet(request)) {
```
```suggestion
        if (GET.matches(request.getMethod())) {
```
조건문을 메서드로 빼는게 꼭 좋은 방법은 아닙니다. 이해하기 힘든 조건을 메서드로 뺐을 때는 좋으나 지금처럼 단순한 경우엔 실제 조건까지 보려면 메서드를 찾아가야하는 수고가 동반되고 클래스 내 메서드들도 비대해지게 되죠. 메서드를 추출해서 발생하는 트레이드 오프를 고려해볼 여지가 있습니다.

```java
for(int i = 0; isUnderTen(i); i++ ) {
```
isUnderTen의 구현이 어떨지는 간단히 유추가 될 겁니다.

Hue9010 commented on src/main/java/wooteco/subway/web/member/interceptor/BearerAuthMemberInterceptor.java:
```diff
@@ -0,0 +1,95 @@
+        request.setAttribute("requestMemberEmail", email);
+
+        if (isGet(request)) {
+            validateEmailEquals(request, email);
```
```java
        String email = jwtTokenProvider.getSubject(bearer);
        request.setAttribute("requestMemberEmail", email);

        if (request.getAttribute("requestMemberEmail").equals(email) == false) {
            throw new NotMatchedEmailIExistInJwtException(email);
        }
```

false가 나올 수 없는 항상 참일 수 밖에는 없는 조건 아닐까요?

Hue9010 commented on src/main/java/wooteco/subway/web/member/argumentresolver/FavoriteMethodArgumentResolver.java:
```diff
@@ -0,0 +1,53 @@
+
+        String favoriteId = (String) webRequest.getAttribute("favoriteId", SCOPE_REQUEST);
+
+        Favorite favoriteById = favoriteService.findFavoriteById(Long.valueOf(favoriteId));
```
지우려고하는 즐겨찾기가 로그인한 회원의 즐겨찾기가 아닌 경우도 존재 할 수 있답니다.

Hue9010 commented on src/main/java/wooteco/subway/web/FavoriteController.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.List;
+
+@RestController
+public class FavoriteController {
```
에러에 대한 처리들이 안보이는데 아래 글들 참고하면 좋을 것 같아요. 방법이 크게 두가지인데 ControllerAdvice(또는 RestControllerAdvice)를 이용해주시면 될 것 같아요.
[스프링부트 : REST 어플리케이션에서 예외처리하기 ](https://springboot.tistory.com/33)
[Spring Handle Exception](https://jaehun2841.github.io/2018/08/30/2018-08-25-spring-mvc-handle-exception/
)
[Spring Guide - Exception 전략](https://cheese10yun.github.io/spring-guide-exception/)

Hue9010 replied:
예외 발생이 필요한 곳들엔 이미 잘 작업해주셔서 핸들러만 만들어주면 금방일 것 같네요. 😄

Ramen6315 replied:
안녕하세요 휴!
리뷰 요청이 늦었습니다 ㅜㅜ...

이번 피드백 수정에 대해서 간략히 comment 남깁니다!

interceptor와 argumentResolver같은경우는 하나의 interceptor와 argumentResolver로 수정했습니다.

이메일 바꿀 수 없는 건 회원 수정시 회원가입때 등록한 로그인아이디 값을 변경 할 수 없어야 한다고 생각했고 미션에선 로그인아이디를 담당하는것이 이메일이라고 생각하여 변경이 불가하도록 설계를했습니다. 그렇기 때문에 서비스코드를 보신것처럼 이메일값에 대한 수정을 구현하지 않았습니다.

첫 구현 과정에서는 resolver와 interceptor에서 많은 것을 처리 하려다 보니 다수의 resolver와 interceptor가 생성되었습니다.
그래서 구현후 resolver와 interceptor의 기능을 줄여 하나의 interceptor와 resolver로 동작하도록 수정했습니다.

이메일 권한과 resolver피드백 주신 부분에 대해선 수정과정 중간에 DM 드릴 수 있었지만
제가 생각하는 불필요한 interceptor, resolver들을 삭제 하다 보니 휴의 피드백 내용을 반영해야 하는 interceptor, resolver들이 많이 삭제되어 중간에 따로 DM을 드리기 보단 추가 피드백 이후 DM을 통해 이야기하는 것이 맞다고 생각했습니다.

늦게 요청 드려 죄송합니다 ㅜㅜㅜ

Hue9010 replied:
ControllerAdvice 잘 적용해주셨네요! 👍 
그리고 일반적으로 생각했을 때 이메일은 변경 불가능한게 맞을 것 같아요. 디테일을 잘 챙겨주셨네요. 😉 
저도 피드백이 늦었네요. 😭 
병렬적으로 진행하시는 것들이 있는걸로 보이는데 너무 조급해하지 마시고 + 복잡하게 생각하지 말고 단순하게 접근해서 생각해보고 진행해보세요.

Hue9010 reviewed (changes requested):
리뷰가 늦었네요. 😢 피드백 남겼습니다. 확인 해주세요.
궁금한 점은 코멘트나 DM으로 문의주세요!!

Ramen6315 committed:
refactor: Controller 수정

 - @RequestMapping 사용

Ramen6315 committed:
refactor: 인증,인가 로직 추상화

- Interceptor -
인증, 인가를 불필요하게 여러 Interceptor로 처리하던 것을 하나의 Interceptor로 단일화시킴

- ArgumentResolver -
처음에 의도 했던 ArgumentResolver보다 많은 책임을 가지고 있었고 그 떄문에 파생된 ArgumentResolver들이 생성됨, 그래서 ArgumentResolver가 가지고 있던 요청 멤버가 그 요청을 처리할 수 있는지 확인하는 절차를 제거히여 의도 했던 ArgumentResolver를 생성하고 단일화 시킴.

- Path Url -
인가가 필요한 요청은 /auth의 경로로 묶어 관리하였다.

- Sql -
즐겨찾기를 의미하는 FAVORITE 테이블의 관계를 파악하기 쉽도록 모든 관계 컬럼들을 해당 테이블의 id로 설정

- 미사용 클래스 삭제

Ramen6315 committed:
refactor: 변경된 인증,인가 로직 테스트 코드 작성

Ramen6315 committed:
feat: 중복 값 저장시 예외처리 기능 구현

Ramen6315 committed:
feat: 중복 이메일 검증 로직 구현

Ramen6315 committed:
feat: globalExceptionHandler 구현

Ramen6315 committed:
refactor : 코드 수정

Ramen6315 committed:
refactor : 테스트 코드 수정

Ramen6315 committed:
refactor : 코드 수정

Hue9010 commented on src/main/java/wooteco/subway/config/DummyDataConfiguration.java:
```diff
@@ -0,0 +1,67 @@
+            lineNumber3.addLineStation(new LineStation(null, station4.getId(), 10, 10));
+            lineNumber3.addLineStation(new LineStation(station4.getId(), station5.getId(), 10, 10));
+
+            lineNumber3 = lineRepository.save(lineNumber3);
```
lineNumber1, lineNumber2, lineNumber3을 save 이후 사용하지 않는데 변수에 다시 담는 이유가 있나요?

Hue9010 commented on src/main/java/wooteco/subway/domain/favorite/FavoriteRepository.java:
```diff
@@ -0,0 +1,19 @@
+    Optional<Favorite> findBySourceAndTarget(@Param("source") String source, @Param("target") String target);
+
+    @Query("SELECT * FROM favorite WHERE member_email = :email")
+    List<Favorite> findByEmail(@Param("email") String email);
```
미사용 코드는 그때 그때 지워주세요!
지금은 불필요한 코드인 걸 알아서 바로 지울 수 있지만 다른 사람 혹은 1년 뒤엔 자신은 지금 지우는 것보다 더 많은 시간을 써야 할 수 있답니다.

Hue9010 commented on src/main/java/wooteco/subway/web/member/interceptor/BearerAuthMemberInterceptor.java:
```diff
@@ -2,37 +2,47 @@
     }
 
+    private void validateToken(String token) {
+        System.out.println(jwtTokenProvider.validateToken(token) + "이거야?" + token);
```
System.out.printX는 개발 도중 디버깅 용도로 사용한 후 최종 코드에선 사용을 지양해주세요.
매번 출력 코드를 추가했다가 지웠다가 번거롭고 실수하기 좋은 이 부분은 나중에 Logger를 배울겁니다. 😉

혹은 다음 사이클은 프로젝트를 먼저 진행하시니 [[Logging] SLF4J를 이용한 Logging](https://gmlwjd9405.github.io/2019/01/04/logging-with-slf4j.html)을 내용을 이해하려고하기보다는 간략히 보시고, 사용법만 익혀보시는 것도 좋을 것 같네요.

Hue9010 commented on src/main/java/wooteco/subway/web/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,64 @@
+    @ExceptionHandler(NotExistStationDataException.class)
+    public ResponseEntity<ErrorResponse> NotExistStationDataException(NotExistStationDataException e) {
+        ErrorResponse errorResponse = new ErrorResponse(e.getMessage());
+        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(errorResponse);
```
UNAUTHORIZED가 아니라 BADREQUEST가 아닐까요?
NotExistXX인 Excpetion이 많은데 NotExistException을 만들고
```
public class NotExistStationDataException extends NotExistException {
...
}
```
위와 같이 만든 다음 ControlelrAdvice에선 공통된 NotExistException을 잡아다가 핸들링 해주면 좋을 것 같네요. 결국 지금 형태도 일종의 중복으로 볼 수 있고 지금과 같은 실수가 끼어들 여지가 커지거든요. 😉

Hue9010 commented on src/main/java/wooteco/subway/web/member/exception/DuplicateException.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.BAD_REQUEST)
```
ControllerAdvice와 동시에 사용할 필요가 있나요?
지금같이 사용한다면 ControllerAdvice와 expcetion 내의 @ResponseStatus 중 어느게 적용될지는 확인이 필요할 것 같테요.

Hue9010 commented on src/main/java/wooteco/subway/domain/favorite/Favorites.java:
```diff
@@ -0,0 +1,42 @@
+        return favorites.stream()
+                .map(Favorite::getId)
+                .collect(Collectors.toList());
+    }
```
미사용 코드는  그때 그때 지워주세요!

Hue9010 replied:
[How to use IntelliJ IDEA to find all unused code?](https://stackoverflow.com/questions/6587729/how-to-use-intellij-idea-to-find-all-unused-code)
다만 위의 경우 코드상에서 직접적으로 사용하지 않는 코드를 다 잡아주기 때문에 지우면 안되는 기본 생성자, getter/setter 등도 잡아줘서 모두 제거 대상이 아니라 그 부분은 구분해주셔야 합니다.

Hue9010 replied:
미사용하는 import 문들도 보이는데 [How to remove unused imports in Intellij IDEA on commit?](https://stackoverflow.com/questions/12585253/how-to-remove-unused-imports-in-intellij-idea-on-commit)에서 나오는 설정 세팅해주는 것도 방법이랍니다. 😉

Hue9010 commented on src/main/java/wooteco/subway/web/FavoriteController.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @GetMapping
+    public ResponseEntity<FavoritesResponse> getFavorite(@LoginMember Member member) {
```
FavoritesResponse를 따로 만들지 않고 List<FavoriteResponse>를 응답해줘도 충분하지 않을까요?

Hue9010 replied:
> FavoritesResponse같은경우 List 형태보단 일급컬렉션으로 다루는게 더 좋아보인다 생각해서 구현했습니다.

정답이 있는 건 아니라 맞다고 생각하시는 방법으로 하시면 됩니다.😉 대신 어떤 이유로 그렇게 생각하시는지도 같이 말해주시는게 좋답니다. 아마 면접때도 이와 같은 질문이 들어 올 수 있는데 보통 어떤 의도를 가지고 만들었는지, 혹은 라면의 생각을 알기 위해서 여쭤보는 거라 이에 대해서도 미리 생각해보시는게 좋답니다.

Hue9010 commented on src/main/java/wooteco/subway/web/member/argumentresolver/LoginMemberMethodArgumentResolver.java:
```diff
@@ -28,14 +30,13 @@ public boolean supportsParameter(MethodParameter parameter) {
+        String email = (String) webRequest.getAttribute("requestMemberEmail", SCOPE_REQUEST);
         try {
             return memberService.findMemberByEmail(email);
         } catch (Exception e) {
```
지금처럼 Exception e로 잡아주시면 db connection 에러 등등 `NotExistMemberDataException`가 아닌 상황에도 `InvalidAuthenticationException `를 발생시킵니다. 그래서 에러 핸들링 할 때는 핸들링하고자 하는 exception을 특정지어 주시는게 예상치 못한 버그 상황을 막을 수 있습니다.
조금더 예를 구체적으로 들어보자면 db와의 connection 이슈로 문제상황이 발생하였을 때 connection 에러가 발생하고 있는 걸 알람을 통해 받으면 바로 connection쪽에 문제가 있다는걸 인지하고 원인파악 및  대응을 할 텐데 지금의 코드로는 connection에 문제가 생겨도 `InvalidAuthenticationException`만 발생하고 있는 걸로 인지해서 이메일을 잘못 입력하는 사용자가 많은 상황으로만 넘어 갈 수 있습니다.(그 숫자가 지나치게 많다면 원인 확인을  시작하겠지만 아무 힌트 없이 원인파악을 해야되는 상황이죠)

그래서 controllerAdvice에서도 RuntimeException, Exception만 핸들링 해주는 건 피해주셔야 합니다. 피해 달라고  했지만 모든 예외를 핸들링 할 수는 없기 때문에 예상못한 예외들에 대한 처리를 위해 존재하긴 해야합니다. (이럴 땐 아마도 status code는 500을 내려주겠죠?)

Hue9010 commented on src/test/java/wooteco/subway/web/favorite/FavoriteControllerIntegrationTest.java:
```diff
@@ -0,0 +1,241 @@
+        assertThat(favoritesResponse.getFavoriteResponses()).hasSize(2);
+    }
+
+    @DisplayName("토큰이 없는 상태에서 즐겨찾기 조회가 실패한다")
```
성공 뿐만 아니라 실패하는 케이스들도 꼼꼼히 잘 해주셨네요! 👍

Hue9010 commented on src/test/java/wooteco/subway/doc/MemberDocumentation.java:
```diff
@@ -1,29 +1,58 @@
+    public static ResultHandler deleteMember() {
+        return document("members/delete",
+                requestHeaders(
+                        headerWithName("Authorization").description("The token for login which is Bearer Type")
```
description도 깔끔히 잘 써주셨네요. 👍 
영어가 충분히 익숙하다면 지금처럼 영어를 사용하는걸 추천드리고 영어로 표현하기 어려운 내용인데 외국인과 같이 일하는 상황이 아니라면 한글을 사용하는 것도 좋습니다.

Hue9010 commented on src/main/java/wooteco/subway/web/FavoriteController.java:
```diff
@@ -0,0 +1,42 @@
+import wooteco.subway.web.member.argumentresolver.annotation.LoginMember;
+
+@RestController
+@RequestMapping("/auth/favorites")
```
`/favorites` -> `/auth/favorites` 로 변경되었는데 화면에서는 기존 `/favorites` 그대로라서 즐겨찾기 추가가 안되네요. /members도 마찬가지인 것 같은데 한번 확인해주세요!
테스트 코드를 많이 짰을 때 거꾸로 실수하기 쉬운 부분인데 실제로도 잘되는지 확인은 해보는게 좋습니다! 백엔드가 주인 과정이지만 이번 과제는 프론트까지 포함하고 있으니 프론트도 잊지 말고 챙겨주세요! 😉

Hue9010 commented on src/main/java/wooteco/subway/config/DummyDataConfiguration.java:
```diff
@@ -0,0 +1,67 @@
+            Member member = new Member("ramen6315@gmail.com", "ramen", "6315");
+
+            favoriteRepository.save(new Favorite(member.getId(), station1.getId(), station2.getId()));
+            favoriteRepository.save(new Favorite(member.getId(), station2.getId(), station3.getId()));
```
Caused by: org.h2.jdbc.JdbcSQLIntegrityConstraintViolationException: NULL not allowed for column "MEMBER_ID";

```sql
create table if not exists FAVORITE
(
    id bigint auto_increment not null,
    member_id bigint not null,
    source_id bigint not null,
    target_id bigint not null,
    primary key(id)
);
```
🤔

지금은 로컬 환경에서만 안뜨지만 이와같은 문제가 특정 환경에 국한되지 않고 발생할 수 있습니다. 테스트 코드가 존재하더라도 최종적으로 간단한 확인 혹은 애플리케이션이 잘 뜨는 지 여부는 확인해주세요!
라면의 경우는 이 곳만 봐도 빠르게 원인을 알 수 있겠지만 다른 사람 혹은 어느정도 시간이 지난뒤의 라면이 로컬에서 작업할 일이 생겨 애플리케이션을 실행시키는데 에러가 발생하면 원인을 찾는데 많은 시간을 쏟아야 할 수도 있답니다.

Hue9010 commented on src/main/java/wooteco/subway/web/member/exception/InvalidAuthenticationException.java:
```diff
@@ -1,11 +1,17 @@
 import org.springframework.http.HttpStatus;
 import org.springframework.web.bind.annotation.ResponseStatus;
 
 @ResponseStatus(HttpStatus.UNAUTHORIZED)
```
왠지 InvalidAuthenticationException엔 한해선 이걸 제거하면 로컬에서는 문제없이 ControllerAdvice에서 잡아서 처리해주는데 테스트에서 못 잡아주고 있네요. 일단 다른 곳들은 ResponseStatus를 제거하고 여긴 놔두도록 하죠.
인터셉터도 적용되는 시점을 고려했을 때는 예외 발생시 ControllerAdvice에서 잡아주는게 맞는데 정확한 원인에 대해선 확인되면 공유드리도록 할게요. 😢

Hue9010 commented on src/main/java/wooteco/subway/web/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,64 @@
+        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(errorResponse);
+    }
+
+//    @ExceptionHandler(Exception.class)
+//    public ResponseEntity<ErrorResponse> remainingException(Exception e) {
+//        ErrorResponse errorResponse = new ErrorResponse(e.getMessage() + "UnHandleException");
+//        return ResponseEntity.badRequest().body(errorResponse);
+//    }
```
슬프지만 서버가 예상치 못한 에러가 발생했을 때 클라도 서버가 아프다는걸 알 수 있게 HttpStatus.INTERNAL_SERVER_ERROR를 내려주세요. 😭

Ramen6315 replied:
안녕하세요 휴!

리뷰가 늦었습니다 ㅜㅜ

제가 커밋을 하나 안올려서 불필요 코드가 같이 있었더라구요 ㅜㅜ 

좀 더 신경썼어야 했는데 ㅜㅜ 죄송합니다!

피드백 주신 부분은 제가 이해한 대로 최대한 수정했습니다!
FavoritesResponse같은경우 List 형태보단 일급컬렉션으로 다루는게 더 좋아보인다 생각해서 구현했습니다.

Hue9010 reviewed (dismissed):
피드백 남겼습니다! 확인해주세요! 
궁금하신 사항은 코멘트나 DM으로 문의주세요~!!

Hue9010 reviewed (changes requested):
피드백 남겼어요 확인해주세요!!
궁금하신 점은 코멘트나 DM으로 문의주시면 됩니다!
(실수로 Appove를 눌러서 취소했는데 혹여나 리뷰가 정상적으로 안보이면 말씀주세요. 🙏 )

Ramen6315 committed:
refactor : 불필요 변수 삭제

Ramen6315 committed:
refactor : 미사용 코드 삭제

Ramen6315 committed:
refactor: NotExistException 구현 및 수정

Ramen6315 committed:
refactor: GlobalExceptionHandler 수정

Ramen6315 committed:
refactor: LoginMemberMethodArgumentResolver Exception 수정 및 LoginMemberControllerTest 수정

Ramen6315 committed:
refactor: ResponseStatus 삭제

Ramen6315 committed:
refactor: uri 수정

Ramen6315 committed:
refactor: 코드 수정

Hue9010 commented on src/main/java/wooteco/subway/web/member/exception/NotExistException.java:
```diff
@@ -0,0 +1,9 @@
+package wooteco.subway.web.member.exception;
+
+public class NotExistException extends IllegalArgumentException {
+    private static final String MESSAGE_TAIL = "에 해당하는 정보가 없습니다.";
+
+    public NotExistException(String message) {
+        super(message + MESSAGE_TAIL);
+    }
```
공통된 메시지를 가지는 것도 좋지만 
`abc@google.com에 해당하는 정보가 없습니다.` 보다는 `존재하지 않는 이메일입니다. [abc@google.com]` 무엇이 잘못 됐는지 직접적으로 알려주는게 더 큰 도움이 될 수 있답니다.
특히나 서버의 로그를 직접 볼 수 있는 담당자를 제외하고는 NotExistFavoriteDataException처럼 `ID = 1에 해당하는 정보가 없습니다.`라는 메시지만 보게된다면 어느 ID인지 알기 힘듭니다.

그래서 각각의 메시지는 별도로 따로 가지는 게 좋다고 봅니다.

Hue9010 reviewed (approved):
고생하셨습니다. 이번 단계는 여기서 머지하도록 할게요!
추가로 궁금하시거나 의문 남는 부분이 있다면 편히 문의 주세요! 😉
